test/mpi: Add communicator refcount test

raffenet opened the pull request:
## Pull Request Description

Add a test that receives a synchronous send *after* the user has freed
the comm in the application. MPI is required to keep the communicator
active because of pending communication. In this case it is needed for
the ACK message to the sender. ch4/am code previously did not pass
this test.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

raffenet commented:
test:mpich/ch4/most

raffenet commented:
Hm, I thought this would fail without #5304, but it seems to be working. I will try on 3.4.x

hzhou commented:
@raffenet Do we still need add this test?

raffenet commented:
Yes, I think this is still good to merge (if it works as expected). I will retest it and verify the results.

raffenet commented:
Just confirmed this test fails with `3.4.x`, so I think it's useful to have in the suite. Re-run again with `main` to get a fresh status.
test:mpich/ch4/ofi

hzhou commented:
> Just confirmed this test fails with `3.4.x`, so I think it's useful to have in the suite. Re-run again with `main` to get a fresh status.

👍 

Can you point to the relevant PRs that fixes this test?

raffenet commented:
I believe it was https://github.com/pmodels/mpich/pull/5304. I'll confirm while Jenkins runs.

hzhou reviewed (approved):
LGTM.

raffenet commented:
Hm, manual testing shows this test passed even without #5304. With #5261, it fails with ch4/posix, but ch4/ofi passes. I this may have been broken and fixed a couple of times as we refactored the generic code.
